apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts

```diff
@@ -0,0 +1,386 @@
+   */
+  private organizeDateMap(
+    rows: Awaited<
+      ReturnType<DelegationPercentageRepository["getDaoMetricsByDateRange"]>
+    >,
+  ): Map<string, DateData> {
+    const dateMap = new Map<string, DateData>();
+
+    rows.forEach((row) => {
+      const dateStr = row.date.toString();
+      const existing = dateMap.get(dateStr) || {};
+
+      if (row.metricType === MetricTypesEnum.DELEGATED_SUPPLY) {
+        existing.delegated = row.high;
+      } else if (row.metricType === MetricTypesEnum.TOTAL_SUPPLY) {
+        existing.total = row.high;
+      }
+
+      existing.daoId = row.daoId;
+      existing.tokenId = row.tokenId;
+
+      dateMap.set(dateStr, existing);
+    });
+
+    return dateMap;
+  }
+
+  /**
+   * Generates a complete date range based on available data
+   * Fills gaps between first and last date with all days
+   * If endDate is not provided, uses current day (today) for forward-fill
+   */
+  private generateDateRange(
+    dateMap: Map<string, DateData>,
+    startDate?: string,
+    endDate?: string,
+    orderDirection: "asc" | "desc" = "asc",
+  ): bigint[] {
+    const allDates: bigint[] = [];
+
+    if (dateMap.size === 0) {
+      return allDates;
+    }
+
+    const datesFromDb = Array.from(dateMap.keys())
+      .map((d) => BigInt(d))
+      .sort((a, b) => Number(a - b));
+
+    const firstDate = startDate ? BigInt(startDate) : datesFromDb[0];
+    const lastDate = endDate ? BigInt(endDate) : getCurrentDayTimestamp();
+
+    if (!firstDate || !lastDate) {
+      return allDates;
+    }
+
+    // Generate all days in range
+    for (
+      let date = firstDate;
+      date <= lastDate;
+      date += BigInt(SECONDS_IN_DAY)
+    ) {
+      allDates.push(date);
+    }
+
+    if (orderDirection === "desc") {
+      allDates.reverse();
+    }
+
+    return allDates;
+  }
+
+  /**
+   * Applies forward-fill logic and calculates delegation percentage
+   * Forward-fill: carries forward the last known value for missing dates
+   */
+  private applyForwardFill(
+    allDates: bigint[],
+    dateMap: Map<string, DateData>,
+    initialValues: { delegated: bigint; total: bigint } = {
+      delegated: 0n,
+      total: 0n,
+    },
+  ): DelegationPercentageItem[] {
+    let lastDelegated = initialValues.delegated;
+    let lastTotal = initialValues.total;
+
+    return allDates.map((date) => {
+      const dateStr = date.toString();
+      const data = dateMap.get(dateStr);
+
+      // Update known values (forward-fill)
+      if (data?.delegated !== undefined) lastDelegated = data.delegated;
+      if (data?.total !== undefined) lastTotal = data.total;
+
+      // Calculate percentage (avoid division by zero)
+      // Returns as string with 2 decimal places (e.g., "11.74" for 11.74%)
+      const percentage =
+        lastTotal > 0n ? Number((lastDelegated * 10000n) / lastTotal) / 100 : 0;
+
+      return {
+        date: dateStr,
+        high: percentage.toFixed(2),
+      };
+    });
```

Comment:
Could something like this work? looks a bit more declarative/stateless-ish (i hate variables with a burning passion)

```suggestion
    const process = (
      dates: bigint[],
      delegated: bigint,
      total: bigint,
    ): DelegationPercentageItem[] => {
      if (dates.length === 0) return [];
      
      const [date, ...rest] = dates;
      const data = dateMap.get(date.toString());
      const newDelegated = data?.delegated ?? delegated;
      const newTotal = data?.total ?? total;
      const percentage = newTotal > 0n ? Number((newDelegated * 10000n) / newTotal) / 100 : 0;
      
      return [
        { date: date.toString(), high: percentage.toFixed(2) },
        ...process(rest, newDelegated, newTotal),
      ];
    };
    
    return process(allDates, initialValues.delegated, initialValues.total);
```